### DISCUSSION TITLE
FATAL: BUG: Struct size mismatch:

### DISCUSSION THREAD

**Discussion by dongho-jung** [bpftrace/bpftrace#2023] [Q&A] [answered]
What I was trying to do is that hook glibc printf and puts and then print its first argument whenever the prove is triggered.
so I tried like this: `sudo bpftrace -e 'BEGIN {@prev = ("","")} u:/usr/lib/libc.so.6:printf,u:/usr/lib/libc.so.6:puts / @prev != (comm, str(arg0)) / { printf("%s %s: %s\n", probe, comm, str(arg0)); @prev = (comm, str(arg0)) }'` the `@prev` is for preventing consecutive duplicate outputs.

However if I run this, I get this error message `FATAL: BUG: Struct size mismatch: expected: 128, real: 80`

what did I do wrong? struct size mismatch...? I have no idea.. thanks..!

**Comment by mmisono** [comment, marked as answer]
The problem is the size of `""` is 64 (STRING_SIZE) but the size of `comm` is 16 (COMM_SIZE). 
https://github.com/iovisor/bpftrace/blob/45d23c82c36d2467e322e96b532edee33720c997/src/ast/passes/semantic_analyser.cpp#L97 https://github.com/iovisor/bpftrace/blob/45d23c82c36d2467e322e96b532edee33720c997/src/ast/passes/semantic_analyser.cpp#L346

(The difference comes from the fact that type of `""` is `char *` but the `comm` is `char[16].)

A workaround is using `comm` instead of `""` in the `BEGIN{}`.
`sudo bpftrace -e 'BEGIN {@prev = (comm,"")} u:/usr/lib/libc.so.6:printf,u:/usr/lib/libc.so.6:puts / @prev != (comm, str(arg0)) / { printf("%s %s: %s\n", probe, comm, str(arg0)); @prev = (comm, str(arg0)) }`

**Reply by dongho-jung** [reply]
wow thanks a lot!  :D
